### ISSUE TITLE
Ability to use Pyodide from a service worker

### ISSUE THREAD

**Issue by simonw** [pyodide/pyodide#2432]
## 🚀 Feature

I'd like to be able to use Pyodide in a service worker, not just in a web worker.

### Motivation

I want to be able to run Python server-side applications (including my [Datasette](https://datasette.io/) application) entirely within the browser.

To do this, I want to register a service worker which intercepts calls to e.g. `/fixtures/facetable` ([this page](https://latest.datasette.io/fixtures/facetable)) and runs my original server-side code using Pyodide.

I've been trying to get this working but then I ran into this error:

![image](https://user-images.githubusercontent.com/9599/166171678-2909fee5-e79d-4a11-b1ba-819059774639.png)

On further research, it looks like `XMLHttpRequest` is supported in web workers but not in service workers: https://stackoverflow.com/questions/37112425/xmlhttprequest-within-service-worker/37114241#37114241

So I think the fix here may be for Pyodide to either drop using `XMLHttpRequest` or to have a fallback for when it's not available. Though there may be plenty of other reasons it doesn't work in a service worker that I've not yet found yet!

### Pitch

I'd like to be able to run Python code in a service worker using Pyodide.

**Comment by simonw** [comment]
It's possible that I don't need this after all - I'm new to service workers, and I'm now wondering if I can achieve what I want using a web worker instead and running additional code using the HTML5 history API whenever a page navigation event is attempted.

**Comment by simonw** [comment]
... on further research I've confirmed that web workers are the way to go for this one. I'll leave this issue open just in case there are other use-cases for which service worker would be appropriate, but I have a strong hunch that this feature isn't actually useful and the issue should be closed.

**Comment by hoodmane** [comment]
I think this is worth fixing. All we need to do is catch and throw away `ImportError` from the failure to load `XMLHttpRequest`. In this case, `open_url` will fail if you try to use it, but that is okay.

**Comment by simonw** [comment]
Here's my write up of what I've built so far: https://simonwillison.net/2022/May/4/datasette-lite/

**Comment by Aierie** [comment]
If someone wants to use pyodide from a background script in a Chrome Extension with manifest v3, they would encounter this problem (+others, since it is a service worker with additional restrictions).

**Comment by rth** [comment]
> So I think the fix here may be for Pyodide to either drop using XMLHttpRequest or to have a fallback for when it's not available. 

I think the [issue is in Emscripten](https://github.com/emscripten-core/emscripten/search?q=XMLHttpRequest) as we don't use `XMLHttpRequest` for anything in Pyodide (aside from `open_url` but that's an unused util function in this case).

**Comment by Aierie** [comment]
I've created a repo to demonstrate changes necessary to use Pyodide in a module type service worker: https://github.com/Aierie/pyodide-for-v3-extension

There are 3 changes necessary to enable use Pyodide in a module type service worker:
1. Polyfill `XMLHttpRequest` - I vendored a no-dependency polyfill
2. Remove the call to `loadScripts` because dynamic  import is not currently allowed
3. Manually import `pyodide.asm.js` since `loadScripts` is no longer doing this

In order to use it without having a `unsafe-eval` CSP (necessary for use in Chrome extension background scripts), you need to remove the Function constructor, which means all callables must be called with `.call` instead.

If someone can outline a good approach for these changes, I'd be happy to work on them.

**Comment by hoodmane** [comment]
1: No change needed in Pyodide for people to polyfill xmlhttprequest =)

2/3: My one concern with the static import is that we need to make sure this doesn't break loading multiple Pyodide instances. I think that it probably will just work. This might also fix some `indexURL`-related problems.

Pretty exciting that we're almost ready to get rid of `loadScripts` entirely. I've been day dreaming about using it to include JavaScript in side modules, but in that case those side modules just wouldn't work without dynamic import.

unsafe-eval CSP: we won't change the API of PyProxy since it'd break a lot of downstream code. We'd have to figure out a way to call it like we currently do. But it might be hard.

**Comment by hoodmane** [comment]
Perhaps we could:

1. Create a subclass of Function: `class My Class extends Function {}`
2. Assign the appropriate methods to `MyClass`
3. Create an empty function with `function(){}`
4. Set the prototype of the function to `MyClass`

**Comment by hoodmane** [comment]
Another possibility is to just use a blank function and move the handling of accessing the desired prototype into the proxy `get` handler.

**Comment by Aierie** [comment]
re: static import, would the approach in https://github.com/pyodide/pyodide/pull/3070 work? I'm concerned that changing to a static import surprises people who don't expect importing `loadPyodide` to be heavy, since it hasn't been heavy before this.

**Comment by hoodmane** [comment]
> In order to use it without having a unsafe-eval CSP (necessary for use in Chrome extension background scripts), you need to remove the Function constructor, which means all callables must be called with PyProxyCallableMethods.call or PyProxyCallableMethods.apply instead.

When I try to set a CSP without unsafe-eval I get an error in Wasm compilation:
```js
wasm streaming compile failed: CompileError: WebAssembly.instantiateStreaming(): Refused to compile or instantiate WebAssembly module because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "default-src http: 'unsafe-inline'"
```
I found this discussion but I don't fully understand the implications:
https://github.com/WebAssembly/content-security-policy/issues/7

**Comment by hoodmane** [comment]
Yes, a new issue for the CSP problem would be good (since your PR #3070 fixes the other points raised here we can close this when #3070 is merged).
